commit 41f45d8df019a6f92719b9ea5a6f7f0b553b3d99
Author: Michael Still <mikal@stillhq.com>
Date:   Tue Jan 28 05:38:23 2025 +1100

    Implement USB controller extra spec for libvirt.

    This is the implementation for the USB controller extra spec as
    desired by the new libvirt spice-direct console mode. USB device
    redirection support is a frequently requested feature for VDI users.

    Note this patch has not yet had the compute protections added.

    Change-Id: I71edd03b5c63a8028c23a746c01c59d303994144

diff --git a/nova/api/validation/extra_specs/hw.py b/nova/api/validation/extra_specs/hw.py
index c50355b..4848d48 100644
--- a/nova/api/validation/extra_specs/hw.py
+++ b/nova/api/validation/extra_specs/hw.py
@@ -552,6 +552,32 @@ feature_flag_validators = [
             'enum': fields.SoundModelType.ALL,
         },
     ),
+    base.ExtraSpecValidator(
+        name='hw:usb_model',
+        description=(
+            'The model of the attached USB controller device. '
+            'Only supported by the libvirt virt driver. '
+            'If unset, no USB controller device is attached.'
+        ),
+        value={
+            'type': str,
+            'description': 'A USB controller model',
+            'enum': fields.USBControllerModelType.ALL,
+        },
+    ),
+    base.ExtraSpecValidator(
+        name='hw:redirected_usb_ports',
+        description=(
+            'The number of redirected USB ports to add to the virtual machine.'
+            'Only supported by the libvirt virt driver. If unset, no '
+            'redirected USB ports are added. The maximum value is 15.'
+        ),
+        value={
+            'type': int,
+            'description': 'The number of USB redirection devices to add',
+            'min': 0,
+        },
+    ),
 ]

 ephemeral_encryption_validators = [
diff --git a/nova/tests/unit/objects/test_image_meta.py b/nova/tests/unit/objects/test_image_meta.py
index 508ed4d..fab3ae8 100644
--- a/nova/tests/unit/objects/test_image_meta.py
+++ b/nova/tests/unit/objects/test_image_meta.py
@@ -569,6 +569,21 @@ class TestImageMetaProps(test.NoDBTestCase):
         primitive = obj.obj_to_primitive('1.27')
         self.assertNotIn('hw_sound_model', primitive['nova_object.data'])

+    def test_obj_make_compatible_usb_model_and_redirected_ports(self):
+        """Check if we pop hw_usb_model and hw_redirected_usb_ports."""
+        obj = objects.ImageMetaProps(
+            hw_usb_model='qemu-xhci',
+            hw_redirected_usb_ports=3
+        )
+        primitive = obj.obj_to_primitive()
+        self.assertIn('hw_usb_model', primitive['nova_object.data'])
+        self.assertIn('hw_redirected_usb_ports', primitive['nova_object.data'])
+
+        primitive = obj.obj_to_primitive('1.27')
+        self.assertNotIn('hw_usb_model', primitive['nova_object.data'])
+        self.assertNotIn(
+            'hw_redirected_usb_ports', primitive['nova_object.data'])
+
     def test_obj_make_compatible_socket_policy(self):
         obj = objects.ImageMetaProps(
             hw_pci_numa_affinity_policy=fields.PCINUMAAffinityPolicy.SOCKET)
diff --git a/nova/virt/hardware.py b/nova/virt/hardware.py
index 01ebaf8..2361bdb 100644
--- a/nova/virt/hardware.py
+++ b/nova/virt/hardware.py
@@ -2936,3 +2936,57 @@ def get_sound_model(
         )

     return model
+
+
+def get_usb_model(
+    flavor: 'objects.Flavor',
+    image_meta: 'objects.ImageMeta',
+) -> ty.Optional[str]:
+    """Get the USB controller model, if any.
+
+    :param flavor: ``nova.objects.Flavor`` instance
+    :param image_meta: ``nova.objects.ImageMeta`` instance
+    :raises: nova.exception.FlavorImageConflict if a value is specified in both
+        the flavor and the image, but the values do not match
+    :raises: nova.exception.Invalid if a value or combination of values is
+        invalid
+    :returns: A string containing the USB controller model, else None.
+    """
+    model = _get_unique_flavor_image_meta('usb_model', flavor, image_meta)
+    if model and model not in fields.USBControllerModelType.ALL:
+        raise exception.Invalid(
+            'Invalid USB controller model %(model)r. '
+            'Allowed values: %(valid)s.'
+            % {
+                'model': model,
+                'valid': ', '.join(fields.USBControllerModelType.ALL)
+            })
+
+    return model
+
+
+def get_redirected_usb_ports(
+    flavor: 'objects.Flavor',
+    image_meta: 'objects.ImageMeta',
+) -> int:
+    """Get the number of redirected USB ports, if any.
+
+    :param flavor: ``nova.objects.Flavor`` instance
+    :param image_meta: ``nova.objects.ImageMeta`` instance
+    :raises: nova.exception.FlavorImageConflict if a value is specified in both
+        the flavor and the image, but the values do not match
+    :raises: nova.exception.Invalid if a value or combination of values is
+        invalid
+    :returns: An integer number of ports, else 0.
+    """
+    count = _get_unique_flavor_image_meta(
+        'redirected_usb_ports', flavor, image_meta)
+    if not count:
+        count = 0
+    elif count > 15:
+        # NOTE(mikal): XHCI controllers only support up to 15 ports. This isn't
+        # documented in the libvirt domain XML documentation at the moment, but
+        # is at https://www.kraxel.org/blog/2018/08/qemu-usb-tips/.
+        raise exception.Invalid('Nova  only supports up to 15 USB ports.')
+
+    return count
diff --git a/nova/virt/libvirt/config.py b/nova/virt/libvirt/config.py
index 2c185c4..d2bcf5f 100644
--- a/nova/virt/libvirt/config.py
+++ b/nova/virt/libvirt/config.py
@@ -4045,3 +4045,19 @@ class LibvirtConfigGuestSound(LibvirtConfigObject):
         meta = self._new_node('sound')
         meta.set('model', str(self.model))
         return meta
+
+
+class LibvirtConfigGuestUSBRedirect(LibvirtConfigObject):
+
+    def __init__(self, **kwargs):
+        super(LibvirtConfigGuestUSBRedirect,
+              self).__init__(root_name='redirdev')
+
+        self.type = 'spicevmc'
+        self.bus = 'usb'
+
+    def format_dom(self):
+        meta = self._new_node('redirdev')
+        meta.set('type', str(self.type))
+        meta.set('bus', str(self.bus))
+        return meta
diff --git a/nova/virt/libvirt/driver.py b/nova/virt/libvirt/driver.py
index fc67f87..9056277 100644
--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -6865,6 +6865,18 @@ class LibvirtDriver(driver.ComputeDriver):
         sound_device = vconfig.LibvirtConfigGuestSound(sound_model)
         guest.add_device(sound_device)

+    def _add_redirected_usb_ports(
+        self,
+        guest: vconfig.LibvirtConfigGuest,
+        flavor: 'objects.Flavor',
+        instance: 'objects.Instance',
+        image_meta: 'objects.ImageMeta',
+    ) -> None:
+        """Add redirected USB ports, if requested."""
+        count = hardware.get_redirected_usb_ports(flavor, image_meta)
+        for i in range(count):
+            guest.add_device(vconfig.LibvirtConfigGuestUSBRedirect())
+
     def _set_qemu_guest_agent(self, guest, flavor, instance, image_meta):
         # Enable qga only if the 'hw_qemu_guest_agent' is equal to yes
         if image_meta.properties.get('hw_qemu_guest_agent', False):
@@ -7378,7 +7390,7 @@ class LibvirtDriver(driver.ComputeDriver):

         return False

-    def _guest_add_usb_root_controller(self, guest, image_meta):
+    def _guest_add_usb_root_controller(self, guest, flavor, image_meta):
         """Add USB root controller, if necessary.

         Note that these are added by default on x86-64. We add the controller
@@ -7387,9 +7399,11 @@ class LibvirtDriver(driver.ComputeDriver):
         """
         usbhost = vconfig.LibvirtConfigGuestUSBHostController()
         usbhost.index = 0
+
         # an unset model means autodetect, while 'none' means don't add a
         # controller (x86 gets one by default)
         usbhost.model = None
+        specified_model = hardware.get_usb_model(flavor, image_meta)
         if not self._guest_needs_usb(guest, image_meta):
             archs = (
                 fields.Architecture.PPC,
@@ -7402,6 +7416,8 @@ class LibvirtDriver(driver.ComputeDriver):
                 # xml, where 'none' adds it but then disables it causing
                 # libvirt errors and the instances not being able to build
                 usbhost.model = None
+            elif specified_model:
+                usbhost.model = specified_model
             else:
                 usbhost.model = 'none'
         guest.add_device(usbhost)
@@ -7566,7 +7582,8 @@ class LibvirtDriver(driver.ComputeDriver):
         if self._guest_needs_pcie(guest):
             self._guest_add_pcie_root_ports(guest)

-        self._guest_add_usb_root_controller(guest, image_meta)
+        self._guest_add_usb_root_controller(guest, flavor, image_meta)
+        self._add_redirected_usb_ports(guest, flavor, instance, image_meta)

         self._guest_add_pci_devices(guest, instance)

